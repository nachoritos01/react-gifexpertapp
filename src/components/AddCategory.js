import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setImputValue] = useState('');
    
    const handleImputChange = (e)=>{
        setImputValue(e.target.value);
        // console.log('handle input change llamado ');

    }

    const handleSubmit = (e)=>{
        e.preventDefault();
        // console.log('hndlessubmit', inputValue);

        if(inputValue.trim().length > 2){

            setCategories(cats=>[ inputValue, ...cats ]);
            setImputValue('');
        }

    }

    return (
        <form onSubmit = {handleSubmit}>
            <p>{ inputValue}</p>

            <input 
                type="text"
                value = {inputValue}
                onChange = {handleImputChange}>

            </input>
        </form>
            
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
