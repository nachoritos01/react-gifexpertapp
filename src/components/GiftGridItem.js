import React from 'react';
import PropTypes from 'prop-types';

export const GiftGridItem = ({title, url}) => {

    return (
        <div className="card animate__animated animate__fadeIn">
            <img alt={title} src={url}/>
            <p>{title}</p>
        </div>
    )
}


GiftGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}
