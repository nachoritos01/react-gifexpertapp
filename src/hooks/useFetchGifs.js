import { useEffect, useState } from "react"
import { getGifts } from "../helpers/getGifs";

/**
 * Retornamos el stado de una funcion con los datos
 * @param {*} category 
 * @returns 
 */
export const useFetchGifs = (category) => {
    //Estado inicial
    const [state, setState] = useState(
        {
            data: [],
            loading: true
        }
    );

    /**
     * Evalia el efecto cuando la categoria cambia y se dispara
     */
    useEffect(()=>{
        //Promesa
        getGifts(category).then(
            //Retorno de los datos
            imgs =>{
                //Nuevo state
                setState({
                    data: imgs,
                    loading: false
                });
            }

        );
    },[category]);

    return state;
    
}
