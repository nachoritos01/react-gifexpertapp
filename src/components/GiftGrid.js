import React from 'react';
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GiftGridItem } from './GiftGridItem';

export const GiftGrid = ({category}) => {

    //Desustructuracion de los datos de retorno, y renombramos el valos de data
    const {data:images, loading} = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>
            { loading && <p className="animate__animated animate__fadeIn">Loading</p>}
            <div className="card-grid">
                {/* Mandamos los valores al componente individualmente cap 75 */}
                { images.map((img)=>( <GiftGridItem key={img.id} {... img}/>))}
            </div>
        </>
    )
};

GiftGrid.propTypes = {
    category: PropTypes.string.isRequired
}