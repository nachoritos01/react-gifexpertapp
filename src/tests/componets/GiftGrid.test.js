import { shallow } from 'enzyme';
import React from 'react'
import { GiftGrid } from '../../components/GiftGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
import '@testing-library/jest-dom';

//Fingir cualquier llamada y suponer o controlar la informacion que va a responder
jest.mock('../../hooks/useFetchGifs');
/**
 * Pruebas del componente gifGrid mock customhook
 */
describe('Pruebas en <GiftGrid/>', ()=>{
    const category = 'Hola mundo';
    // Validamos el estado inicial
    useFetchGifs.mockReturnValue({
        data: [],
        loading: true
    });

    let wrapper = shallow(<GiftGrid category={category}/>);

    test('Debe mostrarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();
        
    });


    test('Dede de mostrar items cuando se cargan imagenes useFetchGifs ', () => {

        const data = [{
            id: '1',
            title: 'test.gif',
            url: 'gift.gif'
        }];
        useFetchGifs.mockReturnValue({
            data,
            loading: false
        });

        const wrapper =  shallow(<GiftGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GiftGridItem').length).toBe(data.length);

        
    });
    
    
});

