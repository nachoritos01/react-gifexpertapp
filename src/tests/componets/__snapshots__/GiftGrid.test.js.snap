// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GiftGrid/> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    Hola mundo
  </h3>
  <p
    className="animate__animated animate__fadeIn"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en <GiftGrid/> Dede de mostrar items cuando se cargan imagenes useFetchGifs  1`] = `
<Fragment>
  <h3>
    Hola mundo
  </h3>
  <div
    className="card-grid"
  >
    <GiftGridItem
      id="1"
      key="1"
      title="test.gif"
      url="gift.gif"
    />
  </div>
</Fragment>
`;
